Factored out some code in the `image-segmentation` pipeline.

Narsil opened the pull request:
# What does this PR do?

Modifies the pipeline's internal a little.

- Hopefully less code, so easier to read.

- Removed the "blank mask" when nothing gets detected. Sending `[]` instead
  is more obvious IMO (and used to be the behavior, so making it non breaking change).
  
- Changed the default of `subtask` to `None` instead of `semantic`. `panoptic` used to be the default, and I merely adopted the same behavior. If a user specifies `subtask=x` we definitely try to use it's task and error out when it's impossible. But when nothing gets passed (should be the most common case), then we try **in order** `"panoptic", "instance", , and "semantic"`. Since roughly they are in decreasing order of expressiveness.

- Added `Detr` as an exception to sending empty lists in the randomly generated
  models (Like maskformer, it *can* return nothing because it has an empty slot
  it seems)
  
- Renamed `task` to `subtask` since `pipeline(task="image-segmentation")` already exists and would be impossible to use from `pipeline` directly. I feel `subtask` is appropriate but welcome any suggestions there.

TODO : figure out the real bug that makes `panoptic` on the tiny detr:

Figured it out. Not sure what the original code was doing but it was outputting two masks for LABEL_215.
What happens is that `pred_masks` is a tensor of shape `(2, H, W)` where `2` is the number of class queries.
Now we only have one mask because we have for all pixels `pred_masks[0, :, :] == pred_masks[1, :, :]`. So when we do
`pred_masks.argmax(dim=0)` to get the class they are associated with, everything gets attributed to class 0, and no pixel to class 1.
Since mask 1 is empty, we exclude it (which is good).

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

Narsil committed:
Factored out some code in the image-segmentation pipeline

Re-enable `small_model_pt`.

Re-enable `small_model_pt`.

Enabling the current test with the current values.

Debugging the values on the CI.

More logs ? Printing doesn't work ?

Using the CI values instead. Seems to be a Pillow sensitivity.

Added a test showcasing that models not supporting some tasks get a
clear error.

Factored out code.

Further factor out.

Fixup.

Bad rebase.

Put `panoptic` before `instance` as it should be a superset.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19727). All of your documentation changes will be reflected on that endpoint.

Narsil commented:
Okay, I've pinpointed out the filter that filters OUT the prediction for panoptic:
(https://huggingface.slack.com/archives/D03VB3U5CVC/p1666108719919289)
36f52e9593
(https://huggingface.slack.com/archives/D03VB3U5CVC/p1666108795507369)
It's this line specifically : https://github.com/huggingface/transformers/pull/19262/files#diff-826e0e0303aa4a12e527884e4dc548fe8debb137905260e42ec8b4331c9c2c63R199
(https://huggingface.slack.com/archives/D03VB3U5CVC/p1666108939394469)
For the random model, the predictions are rather low (score == 0.004) and so >0.5 filter them out, basically loosing the predictions right there.I'm not sure who is correct, but as mentionned earlier, if semantic is saying something, I feel like panoptic without any thresholds should ALWAYS output at least as much.

I think the >0.5 comes from the expectations that sigmoid would have put the range of probs into something normalized and make sure something was in the >0.5 range.I don't really have the time to figure out where the logic fails (if it does) ?

Narsil committed:
Fixing tests.

Narsil commented:
@alaradirik @amyeroberts This PR is ready for review.

Don't hesitate to voice any concern over the changes. This mostly code cleanup. Except for the empty mask being removed (which wasn't the behavior before, and is not tested against).

sgugger reviewed (approved):
LGTM in general, but please wait for @alaradirik review as well since she knows this code more than I do :-)

amyeroberts commented on src/transformers/pipelines/image_segmentation.py:
```diff
@@ -32,7 +32,7 @@ class ImageSegmentationPipeline(Pipeline):
     their classes.
 
-    This image segmentation pipeline can currently be loaded from [`pipeline`] using the following task identifier:
+    This image segmentation pipeline can currently be loaded from [`pipeline`] using the following subtask identifier:
```
I think `image-segmentation` is still the `task` identifier. 
```suggestion
    This image segmentation pipeline can currently be loaded from [`pipeline`] using the following task identifier:
```

Narsil replied:
Too agressive replace. Thanks for the catch !

Narsil replied:
Done.

amyeroberts commented on tests/pipelines/test_pipelines_image_segmentation.py:
```diff
@@ -383,12 +406,11 @@ def test_integration_torch_image_segmentation(self):
     @require_torch
     @slow
     def test_threshold(self):
+        self.maxDiff = None
```
What does this line do?

If we're changing the state of the tester class, we probably want to set it back at the end to keep tests independent.

Narsil replied:
Oops ! forgot to remove (this enables more output in th etests when it fails allowing to show exactly what is different within a nested datastructure)

amyeroberts reviewed (approved):
LGTM! Thanks for the refactor :) 

Two Qs (not for this PR but to understand what's covered) :
 * Do we have the default subtask selection behaviour, panoptic -> instance -> semantic, documented anywhere? 
 * Do we have a test to make sure the expected segmentation task is selected?

Narsil committed:
Adding subtasks tests

+ Fixes `instance` segmentation which was broken due to default and
non kwargs arguments.

Narsil committed:
Fix bad replace.

Narsil commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -1275,12 +1275,13 @@ def post_process_instance_segmentation(
+                pred_labels=pred_labels_item,
+                mask_threshold=mask_threshold,
+                overlap_mask_area_threshold=overlap_mask_area_threshold,
+                label_ids_to_fuse=[],
```
@alaradirik While testing `instance` subtask, I found that the returned masks did not respect `target_size`.
I pinpointed that this call was sending `target_size` to `ids_to_fuse` instead of target_size.

`post_process_instance_segmentation` doesn't take a `label_ids_to_fuse` as a parameter, is that expected ?

Future note:

We can write functions as :
```python
def myfunction(*, arg1, arg2):
   pass
```
And this forbids the caller from calling `myfunction(a, b)` he is forced to use `myfunction(arg1=a, arg2=b)` helping prevent this sort of issue.
I don't think we have to do it, just saying that it's a handy tool.

alaradirik replied:
Thanks for the `target_size` fix @Narsil! And yes, `post_process_instance_segmentation` doesn't and shouldn't take a `label_ids_to_fuse` as a parameter.

I'm more in favor of using named arguments but both are fine with me really.

alaradirik reviewed (approved):
All looks good to me! Thanks for the fix and code clean up.
